### PR TITLE
Improve websocket reconnect robustness

### PR DESCRIPTION
**Pull request by mcm001**
Refactors curst stack overflow code to Chris's stripped down NT4 client. Much better on Firefox, at least for me

### CODE REVIEW

File: photon-client/src/plugins/ReconnectingWebsocket.js [hunk 1]
```diff
@@ -0,0 +1,79 @@
...
+        console.log("[NT4] Websocket error - " + e.toString());
+        this.ws.close();
+    }
+
+    ws_onMessage(e) {
+       this.onDataIn(e);
+    }
+
+    ws_connect() {
+        this.clientIdx = Math.floor(Math.random() * 99999999); //Not great, but using it for now
```
Comment (gerth2) [inline photon-client/src/plugins/ReconnectingWebsocket.js:66]: This is NT4 client specific and should be able to be removed.

File: photon-client/src/plugins/ReconnectingWebsocket.js [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
+        setTimeout(this.ws_connect.bind(this), 500);
+
+        if (!e.wasClean) {
+            console.error('Socket encountered error!');
+        }
+
+    }
+
+    ws_onError(e) {
+        console.log("[NT4] Websocket error - " + e.toString());
```
Comment (gerth2) [inline photon-client/src/plugins/ReconnectingWebsocket.js:57]: [NT4] -> [WebSocket]

**Review by gerth2 — DISMISSED** [review]
Couple minor tweaks

**Comment by mdurrani808** [comment]
Still getting spams on page refresh but feels a lot snappier on my end so.

[2023-01-05 16:10:03] [General - NTDataPublisher] [DEBUG] Pipeline index is already 0
[2023-01-05 16:10:03] [General - NTDataPublisher] [DEBUG] Driver mode is already false
[2023-01-05 16:10:04] [General - ConfigManager] [DEBUG] Saving to disk...
[2023-01-05 16:10:04] [General - ConfigManager] [INFO] Settings saved!
[2023-01-05 16:10:06] [WebServer - Server] [DEBUG] Handled HTTP GET request from [0:0:0:0:0:0:0:1] in 18.5918ms
[2023-01-05 16:10:06] [WebServer - Server] [DEBUG] Handled HTTP GET request from [0:0:0:0:0:0:0:1] in 42.0365ms
[2023-01-05 16:10:06] [WebServer - DataSocketHandler] [INFO] Closing websocket connection from /0:0:0:0:0:0:0:1:52640 for reason: Connection closed by client
[2023-01-05 16:10:06] [WebServer - Server] [DEBUG] Handled HTTP GET request from [0:0:0:0:0:0:0:1] in 6.931ms
[2023-01-05 16:10:06] [WebServer - Server] [DEBUG] Handled HTTP GET request from [0:0:0:0:0:0:0:1] in 3.8062ms
[2023-01-05 16:10:06] [WebServer - Server] [DEBUG] Handled HTTP GET request from [0:0:0:0:0:0:0:1] in 6.135ms
[2023-01-05 16:10:06] [WebServer - Server] [DEBUG] Handled HTTP GET request from [0:0:0:0:0:0:0:1] in 2.3869ms
[2023-01-05 16:10:06] [WebServer - Server] [DEBUG] Handled HTTP GET request from [0:0:0:0:0:0:0:1] in 1.4814ms
[2023-01-05 16:10:06] [WebServer - Server] [DEBUG] Handled HTTP GET request from [0:0:0:0:0:0:0:1] in 1.1591ms
[2023-01-05 16:10:06] [WebServer - Server] [DEBUG] Handled HTTP GET request from [0:0:0:0:0:0:0:1] in 1.4282ms
[2023-01-05 16:10:06] [WebServer - Server] [DEBUG] Handled HTTP GET request from [0:0:0:0:0:0:0:1] in 1.4598ms
[2023-01-05 16:10:06] [WebServer - DataSocketHandler] [INFO] New websocket connection from /0:0:0:0:0:0:0:1:52655
[2023-01-05 16:10:06] [WebServer - Server] [DEBUG] Handled HTTP GET request from [0:0:0:0:0:0:0:1] in 3.1068ms
[2023-01-05 16:10:06] [WebServer - Server] [DEBUG] Handled HTTP GET request from [0:0:0:0:0:0:0:1] in 3.2199ms
[2023-01-05 16:10:07] [WebServer - Server] [DEBUG] Handled HTTP GET request from [0:0:0:0:0:0:0:1] in 2.7888ms
[2023-01-05 16:10:07] [WebServer - Server] [DEBUG] Handled HTTP GET request from [0:0:0:0:0:0:0:1] in 2.204ms

**Comment by mcm001** [comment]
Handled HTTP GET request is normal, that's requests for html/css/js files. We care about things that say "websocket connection"

File: photon-client/src/views/PipelineView.vue [hunk 3]
```diff
@@ -446,7 +446,13 @@ export default {
...
       get() {
         // For now we only show the FPS is too low warning when GPU acceleration is enabled, because we don't really trust the presented video modes otherwise
-        return this.$store.state.pipelineResults.fps - this.$store.getters.currentVideoFormat.fps < -5 && this.$store.state.pipelineResults.fps !== 0 && !this.$store.getters.isDriverMode && this.$store.state.settings.general.gpuAcceleration;
+        const currFPS = this.$store.state.pipelineResults.fps;
+        const targetFPS = this.$store.getters.currentVideoFormat.fps;
+        const driverMode = this.$store.getters.isDriverMode;
+        const gpuAccel = this.$store.state.settings.general.gpuAcceleration === true;
+        const isReflective = this.$store.getters.pipelineType === 2;
+
+        return (currFPS - targetFPS) < -5 && this.$store.state.pipelineResults.fps !== 0 && !driverMode && gpuAccel && isReflective;
```
Comment (pietroglyph) [inline photon-client/src/views/PipelineView.vue:455]: Yeah this was messy. Good choice.

File: photon-client/src/App.vue [hunk 4]
```diff
@@ -210,31 +208,21 @@ export default {
...
+      (event) => {
+        event; this.$store.commit("backendConnected", true);
+        this.$store.state.connectedCallbacks.forEach(it => it());
+       },
 
-        this.$store.commit("backendConnected", true)
-        this.$store.state.connectedCallbacks.forEach(it => it())
-      };
+      // on disconnect
+      (event) => { event; this.$store.commit("backendConnected", false) }
```
Comment (pietroglyph) [inline photon-client/src/App.vue:220]: Why the `event;` expression?

File: photon-client/src/plugins/ReconnectingWebsocket.js [hunk 5]
```diff
@@ -0,0 +1,74 @@
...
+        this.ws = null;
+        this.serverConnectionActive = false;
+
+        // User connection-closed hook
+        this.onDisconnect();
+
+        console.log('[WebSocket] Socket is closed. Reconnect will be attempted in 0.5 second.', e.reason);
+        setTimeout(this.ws_connect.bind(this), 500);
+
+        if (!e.wasClean) {
```
Comment (pietroglyph) [inline photon-client/src/plugins/ReconnectingWebsocket.js:47]: This isn't set anywhere.

File: photon-client/src/App.vue (hunk 4 above)
Comment (pietroglyph) [inline photon-client/src/App.vue:220]: Why the `event;` expression?

**Review by pietroglyph — APPROVED** [review]
At a glance nothing seems problematic and the new `ReconnectingWebsocket` seems to do the things we need to do/did before.

**Pull request merged**